eyeshade/lib/transaction.js

```diff
@@ -1,27 +1,98 @@
 const BigNumber = require('bignumber.js')
 const getPublisherProps = require('bat-publisher').getPublisherProps
 const uuidv5 = require('uuid/v5')
+const uuid = require('uuid')
 const {
   createdTimestamp,
   normalizeChannel
 } = require('bat-utils/lib/extras-utils')
-
 const getYoutubeChannelId = require('bat-utils/lib/youtube')
 
+const knownChains = {
+  ETH: 'ethereum',
+  BTC: 'bitcoin',
+  LTC: 'litecoin'
+}
 const SETTLEMENT_NAMESPACE = {
   'contribution': '4208cdfc-26f3-44a2-9f9d-1f6657001706',
   'referral': '7fda9071-4f0d-4fe6-b3ac-b1c484d5601a',
   'manual': 'a7cb6b9e-b0b4-4c40-85bf-27a0172d4353'
 }
+const insertTransaction = `
+INSERT INTO transactions (
+  id,
+  created_at,
+  description,
+  transaction_type,
+  document_id,
+  from_account,
+  from_account_type,
+  to_account,
+  to_account_type,
+  amount,
+  settlement_currency,
+  settlement_amount,
+  channel
+)
+VALUES (
+  $1, to_timestamp($2), $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13
+)
+RETURNING *;
+`
 
 module.exports = {
+  knownChains: Object.assign({}, knownChains),
+  insertFromChain,
   insertFromSettlement,
   insertFromVoting,
   insertFromReferrals,
   updateBalances,
   insertFromAd
 }
 
+async function insertFromChain (runtime, client, chainDoc = {}) {
+  let {
+    id,
+    amount,
+    chain,
+    owner,
+    createdAt,
+    address
+  } = chainDoc
+
+  // using this so that we can pass in ticker / full name for now
+  chain = knownChains[chain] || chain
+
+  if (!amount) {
+    throw new Error('Missing amount field')
+  }
+
+  amount = new BigNumber(amount)
+  if (amount.lessThanOrEqualTo(0)) {
+    return [] // skip because we don't track tx's that don't have an accounting purpose
+  }
+
+  const {
+    rows
+  } = await client.query(insertTransaction, [
+    // surveyorId and channel pair should be unique
+    uuid.v4(),
```

Comment:
what do you think about using uuid.v5 based on the on chain hash?

Reply:
yeah, sure. that makes sense. so that we can't have 2 of the same tx ids